### PR TITLE
Encoder Global Cache Manager

### PR DESCRIPTION
**Pull request by liusy58**
## Motivation

This PR introduces a multi-level multimodal embedding cache powered by Mooncake distributed store, enabling cross-instance sharing of Vision Transformer (ViT) embeddings to avoid redundant GPU computation for previously processed images.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/disaggregation/encode_server.py [hunk 1]
```diff
@@ -290,6 +301,159 @@ async def _flatten_and_load_images(self, mm_items):
...
+        if hit_indices:
+            hit_hashes = [image_hashes[i] for i in hit_indices]
+            hit_tokens = [self.get_num_tokens(grid_thw[i]) for i in hit_indices]
+            self.mm_global_cache.prefetch(req_id, hit_hashes, hit_tokens)
+
+        new_slices = await gpu_task if gpu_task else []
+
+        if hit_indices:
+            while not self.mm_global_cache.check_prefetch_progress(req_id):
+                await asyncio.sleep(0.001)
```
Comment (ZhengWG) [inline python/sglang/srt/disaggregation/encode_server.py]: Consider using `asyncio.Event` here instead.

File: python/sglang/srt/managers/embedding_cache_controller.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                op = self.ongoing_prefetch[req_id]
+                if op.is_finished:
+                    local_ready = op.success
+
+        if self.tp_world_size > 1:
+            ready_tensor = torch.tensor(
+                [1 if local_ready else 0], dtype=torch.int, device="cpu"
+            )
+            torch.distributed.all_reduce(
+                ready_tensor, op=torch.distributed.ReduceOp.MIN, group=self.tp_group
```
[the remark this answers is not in the document]
Comment (ZhengWG) [reply python/sglang/srt/managers/embedding_cache_controller.py]: the comment is valid: The tp_group attribute is referenced but never initialized in the class.

File: python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+                    # Allocate from this block
+                    remaining_size = block_size - size_bytes
+                    if remaining_size > 0:
+                        self.free_blocks[i] = (offset + size_bytes, remaining_size)
+                    else:
+                        self.free_blocks.pop(i)
+                    return offset
+            return None
+
+    def free(self, offset: int, size_bytes: int):
```
Comment (ZhengWG) [inline python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py:44]: The `free()` function appears to be unused.

File: python/sglang/srt/managers/embedding_cache_controller.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+
+class EmbeddingCacheController:
+    def __init__(self, tp_rank, tp_size, max_pool_size_gb=4.0, hidden_dim=1024):
+        self.tp_world_size = tp_size
+        self.hidden_dim = hidden_dim
+        self.element_size = torch.float32.itemsize
+
+        # 1. Mooncake Backend & Pinned Buffer
+        mooncake_config = MooncakeStoreConfig.load_from_env()
```
Comment (ZhengWG) [inline python/sglang/srt/managers/embedding_cache_controller.py]: Consider using `StorageBackendFactory.create_backend()` here to support other storage backends in the future.

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import logging
+from typing import Any, List
+
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+class MooncakeEmbeddingStore:
```
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py]: Can we directly inherit from MooncakeStore on the Hicache side?

Comment (liusy58) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py]: I will do it.

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py:1]: I think we should abstract a Base Class named Mooncake Store, and MooncakeEmbeddingStore inherited from it. They can reuse methods such as register_buffer

File: python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py
Comment (hzh0425) [inline python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py:1]: pls move this file to 'mem_cache/storage'

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        self,
+        storage_config: Any = None,
+    ):
+        super().__init__()
+
+        MooncakeDistributedStore = self._import_mooncake_store()
+        self.store = MooncakeDistributedStore()
+        self.config = self._load_config(storage_config)
+        self.store = MooncakeDistributedStore()
+        ret_code = self.store.setup(
```
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py:19]: Why do we need to init it twice?

Comment (liusy58) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py:19]: fixed

File: python/sglang/srt/server_args.py [hunk 7]
```diff
@@ -4653,6 +4654,13 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="Enable prefix multimodal cache. Currently only supports mm-only.",
         )
 
+        parser.add_argument(
+            "--enable-mm-global-cache",
+            action="store_true",
+            default=ServerArgs.enable_mm_global_cache,
+            help="Enable.",
```
[the remark this answers is not in the document]
Comment (stmatengss) [reply python/sglang/srt/server_args.py]: Consider this comment

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        if put_keys:
+            results = self.store.batch_put_from(put_keys, put_ptrs, put_sizes)
+            for i, res in enumerate(results):
+                success_map[indices[i]] = res == 0
+        return success_map
+
+    def batch_is_exist(self, hashes: List[str]) -> List[bool]:
+        keys = [self.get_key(h) for h in hashes]
+        results = self.store.batch_is_exist(keys)
+        return [res == 1 for res in results]
```
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_embedding_store.py:68]: Can we reuse these functions from mooncake store in HiCache side?

File: python/sglang/srt/disaggregation/encode_server.py [hunk 9]
```diff
@@ -308,6 +320,165 @@ async def _flatten_and_load_images(self, mm_items):
...
```
Comment (stmatengss) [inline python/sglang/srt/disaggregation/encode_server.py]: Why do we need to sleep for 0.01s.  Is it for waiting IO?

**Review by xiezhq-hermann — APPROVED** [review]

**Review by hzh0425 — APPROVED** [review]

**Review by stmatengss — APPROVED** [review]

**Comment by stmatengss** [comment]
After https://github.com/sgl-project/sglang/pull/18460 is merged, Transfer engine can be reused for Mooncake store.. The current version should initialize the mooncake store without reuse.

File: python/sglang/srt/server_args.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShangmingCai) [reply python/sglang/srt/server_args.py]: Yeah. The help info is not complete, it should be fixed before getting merged.

File: python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+        # 3. Task Tracking
+        self.ongoing_prefetch = {}  # {req_id: EmbeddingPrefetchOperation}
+        self.prefetch_queue = Queue()
+        self.insert_queue = Queue()
+
+        self.lock = threading.Lock()
+        self.stop_event = threading.Event()
+        self.io_thread = threading.Thread(target=self._io_loop, daemon=True)
+        self.io_thread.start()
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py:128]: `EmbeddingCacheController` doesn't seem to have `tp_group`, so which tp_group exactly does `check_prefetch_progress` use?

File: python/sglang/srt/disaggregation/encode_server.py [hunk 11]
```diff
@@ -308,6 +320,165 @@ async def _flatten_and_load_images(self, mm_items):
...
+
+        cached_slices = (
+            self.mm_global_cache.get_embeddings([image_hashes[i] for i in hit_indices])
+            if hit_indices
+            else []
+        )
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/encode_server.py]: Will prefetch fail? Do we need to handle the case where the hash is hit, but the prefetch fails?

File: python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+
+import asyncio
+import logging
+import threading
+from typing import List, Optional
+
+import torch
+
+logger = logging.getLogger(__name__)
+
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py]: These imports are dirty, please combine and fix them.

Comment (liusy58) [reply python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py]: done

File: python/sglang/srt/disaggregation/encode_server.py (hunk 11 above)
Comment (liusy58) [reply python/sglang/srt/disaggregation/encode_server.py]: I have added some code to handle this. PTAL.

File: python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py (hunk 10 above)
Comment (liusy58) [reply python/sglang/srt/mem_cache/storage/mooncake_store/embedding_cache_controller.py:128]: Added now.
